File: ruby-resend-examples/examples/metrics.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# Email Metrics
#
# Demonstrates retrieving account-level email metrics, broken down
# by period and broadcast.
#
# Usage:
#   ruby examples/metrics.rb [broadcast_id]
#
# @see https://resend.com/docs/api-reference/emails/get-metrics

require "bundler/setup"
require "resend"
require "dotenv/load"

Resend.api_key = ENV.fetch("RESEND_API_KEY")

puts "=== Email Metrics ===\n\n"

# Totals for the default range (last 6 days), no breakdown
puts "Fetching totals..."
totals = Resend::Emails.metrics
puts "  Sent: #{totals[:totals]["sent"]}"
puts "  Delivered: #{totals[:totals]["delivered"]}"
puts "  Bounced: #{totals[:totals]["bounced"]}"
puts

broadcast_id = ARGV[0]

if broadcast_id
  puts "Fetching metrics for broadcast #{broadcast_id}..."
  params = {
    start_date: "2026-07-01",
    end_date: "2026-07-08",
    dimensions: ["period", "broadcast"],
    broadcast_id: [broadcast_id]
  }

  metrics = Resend::Emails.metrics(params)

  if metrics[:data]&.any?
    puts "Breakdown by period, broadcast:"
    metrics[:data].each do |row|
      puts "  Period: #{row["period"]}"
      puts "  Broadcast: #{row["broadcast_name"]}" if row["broadcast_name"]
      puts "  Sent: #{row["sent"] || 0}"
      puts
    end
  end
end

File: python-resend-examples/examples/metrics.py
#!/usr/bin/env python3
"""
Email Metrics

Demonstrates retrieving account-level email metrics using the Resend API.

Requires resend>=2.40.0.

Usage:
    python examples/metrics.py

See: https://resend.com/docs/api-reference/emails/get-metrics
"""

import os
import resend
from dotenv import load_dotenv

load_dotenv()

resend.api_key = os.environ["RESEND_API_KEY"]

print("=== Email Metrics ===\n")

# Totals for the default range (last 6 days), no breakdown
print("Fetching totals...")
totals = resend.Emails.metrics()
print(f"  Sent: {totals['totals'].get('sent', 0)}")
print(f"  Delivered: {totals['totals'].get('delivered', 0)}")
print(f"  Bounced: {totals['totals'].get('bounced', 0)}")
print()

# Broken down by period and broadcast, filtered to a specific broadcast
broadcast_id = "5a5a3b1e-3b1a-4b1a-8b1a-3b1a4b1a8b1a"
print(f"Fetching metrics for broadcast {broadcast_id}...")
params: resend.Emails.MetricsParams = {
    "start_date": "2026-07-01",
    "end_date": "2026-07-08",
    "dimensions": ["period", "broadcast"],
    "broadcast_id": [broadcast_id],
}
metrics = resend.Emails.metrics(params)
for row in metrics.get("data", []):
    print(f"  {row['period']}: {row.get('sent', 0)} sent, {row.get('delivered', 0)} delivered")

# Example: metrics broken down by domain, mutually exclusive with email/broadcast
# resend.Emails.metrics({
#     "dimensions": ["domain"],
#     "domain_id": ["d91cd9bd-1176-4f47-2a4b-fce2d5399cbf"],
# })
